File: OrangeJuice.Server.Core/Security/AuthTokenFactory.cs
using System;
using System.Net.Http;
using System.Net.Http.Formatting;
using System.Threading.Tasks;

using Factory;

using OrangeJuice.Server.Configuration;
using OrangeJuice.Server.Data.Models;

namespace OrangeJuice.Server.Security
{
	public sealed class AuthTokenFactory : IFactory<Task<AuthToken>, AuthToken>
	{
		private readonly AuthOptions _authOptions;

		public AuthTokenFactory(AuthOptions authOptions)
		{
			_authOptions = authOptions;
		}

		public async Task<AuthToken> Create(AuthToken authorizationToken)
		{
			var httpClient = new HttpClient { BaseAddress = new Uri("https://orangejuice.auth0.com") };
			httpClient.DefaultRequestHeaders.Add("Accept", "application/json");

			var request = new
			{
				client_id = _authOptions.Audience,
				access_token = authorizationToken.AccessToken,
				connection = "google-oauth2",
				scope = "openid"
			};

			var response = await httpClient.PostAsync("oauth/access_token", request, new JsonMediaTypeFormatter());
			response.EnsureSuccessStatusCode();

			var formatters = new[]
			{
				new JsonMediaTypeFormatter
				{
					SerializerSettings =
					{
						ContractResolver = new UnderscoreMappingResolver()
					}
				}
			};
			return await response.Content.ReadAsAsync<AuthToken>(formatters);
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Security/AuthTokenFactory.fs
namespace OrangeJuice.Server.FSharp.Security

open System
open System.Net.Http
open System.Net.Http.Formatting
open System.Runtime.Serialization
open System.Text.RegularExpressions
open System.Threading.Tasks

open OrangeJuice.Server.Configuration
open OrangeJuice.Server.Data.Models

[<DataContract>]
type AuthTokenRequest = {
    [<field: DataMember(Name="client_id")>]
    client_id : string
    
    [<field: DataMember(Name="access_token")>]
    access_token : string
    
    [<field: DataMember(Name="connection")>]
    connection : string
    
    [<field: DataMember(Name="scope")>]
    scope : string
}

type UnderscoreMappingResolver() =
    inherit Newtonsoft.Json.Serialization.DefaultContractResolver()

    override this.ResolvePropertyName(propertyName : string) : string =
        Regex.Replace(propertyName, "([A-Z])([A-Z][a-z])|([a-z0-9])([A-Z])", "$1$3_$2$4", RegexOptions.Compiled).ToLower()

type AuthTokenFactory(authOptions : AuthOptions) =
    interface Factory.IFactory<Task<AuthToken>, AuthToken> with
        member this.Create(authorizationToken : AuthToken) : Task<AuthToken> =
            let task = async {
                let httpClient = new HttpClient (
                    BaseAddress = new Uri("https://orangejuice.auth0.com"))
                httpClient.DefaultRequestHeaders.Add("Accept", "application/json") |> ignore

                let request = {
                    client_id = authOptions.Audience;
                    access_token = authorizationToken.AccessToken;
                    connection = "google-oauth2";
                    scope = "openid" }

                let! response = httpClient.PostAsync("oauth/access_token", request, new JsonMediaTypeFormatter()) |> Async.AwaitTask
                response.EnsureSuccessStatusCode() |> ignore
                
                let formatter = new JsonMediaTypeFormatter()
                formatter.SerializerSettings.ContractResolver <- new UnderscoreMappingResolver()

                let! content = response.Content.ReadAsAsync<AuthToken>([| formatter :> MediaTypeFormatter |]) |> Async.AwaitTask
                return content
            }
            task |> Async.StartAsTask